src/HierarchicalCachePoolTrait.php

```diff
@@ -0,0 +1,101 @@
+<?php
+
+namespace Cache\Hierarchy;
+
+use Cache\Adapter\Common\Exception\InvalidArgumentException;
+use Cache\Taggable\TaggablePoolInterface;
+
+/**
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+trait HierarchicalCachePoolTrait
+{
+    /**
+     * A temporary cache for keys
+     *
+     * @type array
+     */
+    private $keyCache = [];
+
+    /**
+     * Get a value form the store. This must not be an PoolItemInterface.
+     *
+     * @param string $key
+     *
+     * @return string|null
+     */
+    abstract protected function getValueForStore($key);
+
+
+    /**
+     * Get a key to use with the hierarchy. If the key does not start with HierarchicalPoolInterface::SEPARATOR
+     * this will return an unalterered key. This function supports a tagged key. Ie "foo:bar".
+     *
+     * @param string $key The original key
+     * @param string &$pathKey A cache key for the path. If this key is changed everything beyond that path is changed.
+     *
+     * @return string
+     */
+    protected function getHierarchyKey($key, &$pathKey = null)
+    {
+        if (!$this->isHierarchyKey($key)) {
+            return $key;
+        }
+
+        $key = $this->explodeKey($key);
+
+        $keyString = '';
+        $pathKey = 'path:';
+        foreach ($key as $name) {
+            $keyString .= $name;
+            $pathKey = 'path:' . $keyString;
+
+            if (isset($this->keyCache[$pathKey])) {
+                $index = $this->keyCache[$pathKey];
+            } else {
+                $index = $this->getValueForStore($pathKey);
+                $this->keyCache[$pathKey] = $index;
+            }
+
+            $keyString .= ':' . $index . ':';
```

Comment:
Yes. Consider the key: `|foo|bar|baz`
After the first loop it will the `$keyString` will be: `foo:[foo_index]:`. After next iteration: `foo:[foo_index]:bar:[bar_index]`. 

Note that with a clean cache the `$keyString` will be `foo::bar::baz:` and after clearing `|foo|bar` the `$keyString` will be `foo::bar:1:baz:`.